pkg/kubelet/container/cache.go

```diff
@@ -0,0 +1,204 @@
+			// The global timestamp is newer than minTime, return status.
+			return &data{status: &PodStatus{ID: id}, err: nil}
+		}
+	} else {
+		// Status is cached, return status if either of the following is true.
+		//   * status was modified after minTime
+		//   * the global timestamp of the cache is newer than minTime.
+		if d.modified.After(minTime) || c.timestamp.After(minTime) {
+			return d
+		}
+	}
+
+	// The pod status is not ready.
+	return nil
+}
+
+// Get returns the PodStatus for the pod; callers are expected not to
+// modify the objects returned.
+func (c *cache) Get(id types.UID) (*PodStatus, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	d := c.get(id)
+	return d.status, d.err
+}
+
+func (c *cache) get(id types.UID) *data {
+	d, ok := c.pods[id]
+	if !ok {
+		// A cache miss means that there are no states regarding the pod when
+		// querying the container runtime; return an empty PodStatus.
+		return &data{status: &PodStatus{ID: id}, err: nil}
+	}
+	return d
+}
+
+// Set sets the PodStatus for the pod.
+func (c *cache) Set(id types.UID, status *PodStatus, err error, timestamp time.Time) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	defer c.notify(id, timestamp)
+	if err == nil && len(status.ContainerStatuses) == 0 {
+		// Remove if status is empty. This allows the cache to reclaims the
+		// unused entries without needing explicit garbage collection methods.
+		delete(c.pods, id)
+		return
+	}
+	c.pods[id] = &data{status: status, err: err, modified: timestamp}
+}
+
+// SetTimestamp modifies the global timestamp of the cache.
+func (c *cache) SetGlobalTimestamp(timestamp time.Time) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.timestamp = &timestamp
+	c.notifyAll()
+}
+
+func nonblockingSend(ch chan *data, d *data) {
+	select {
+	case ch <- d:
+	default:
+		// Don't block.
+	}
+}
+
+// notify sends notifications for pod with the given id, if the requirements
+// are met. Note that the caller should acquire the lock.
+func (c *cache) notify(id types.UID, timestamp time.Time) {
+	list, ok := c.subscribers[id]
+	if !ok {
+		// No one to notify.
+		return
+	}
+	newList := []*subRecord{}
+	for i, r := range list {
+		if timestamp.Before(r.time) {
+			// Doesn't meet the time requirement; keep the record.
+			newList = append(newList, list[i])
+			continue
+		}
+		nonblockingSend(r.ch, c.get(id))
+		close(r.ch)
+	}
+	if len(newList) == 0 {
+		delete(c.subscribers, id)
+	} else {
+		c.subscribers[id] = newList
+	}
+}
+
+func (c *cache) notifyAll() {
+	for id := range c.subscribers {
+		c.notify(id, *c.timestamp)
+	}
+}
+
+func (c *cache) subscribe(id types.UID, timestamp time.Time, ch chan *data) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	d := c.getIfNewerThan(id, timestamp)
+	if d != nil {
+		nonblockingSend(ch, d)
+		return
+	}
+	// Add the subscription record.
+	if _, ok := c.subscribers[id]; !ok {
+		c.subscribers[id] = []*subRecord{}
+	}
+	c.subscribers[id] = append(c.subscribers[id], &subRecord{time: timestamp, ch: ch})
```

Comment:
For now, we don't have multiple subscribers. However, it seems dumb to have a GetPodIfNewer() method in a cache that can not support multiple callers for a single pod.